File: controllers/proofIdController.js
const ProofId = require('../model/proofid'); // Assuming the schema file is named proofId.js
const AuditLog = require('../model/Auditlogs'); // Import the Audit Log model

// Create a new proof ID
exports.createProofId = async (req, res) => {
  try {
    const userId = req.user._id;
    req.body.userId = userId;
    const proofId = new ProofId(req.body);
    await proofId.save();

    // Create an audit log entry for the proof ID creation
    await AuditLog.create({
      userId: req.user._id,
      action: 'create',
      entity: 'ProofId',
      entityId: proofId._id,
      newValue: proofId,
      ipAddress: req.ip,
      userAgent: req.headers['user-agent']
    });

    res.status(201).send(proofId);
  } catch (error) {
    res.status(400).send(error);
  }
};

// Get all proof IDs
exports.getAllProofIds = async (req, res) => {
  try {
    const proofIds = await ProofId.find({}).populate({
      path: 'userId',
      select: 'name email'
    });

    // Create an audit log entry for retrieving proof IDs
    await AuditLog.create({
      userId: req.user._id,
      action: 'view',
      entity: 'ProofId',
      entityId: null, // No specific entity ID for this action
      newValue: proofIds,
      ipAddress: req.ip,
      userAgent: req.headers['user-agent']
    });

    res.status(200).json({ proofIds });
  } catch (error) {
    res.status(500).send(error);
  }
};

// Get a proof ID by ID
exports.getProofIdById = async (req, res) => {
  const _id = req.params.id;
  try {
    const proofId = await ProofId.findById(_id);
    if (!proofId) {
      return res.status(404).send();
    }

    // Create an audit log entry for retrieving a specific proof ID
    await AuditLog.create({
      userId: req.user._id,
      action: 'view',
      entity: 'ProofId',
      entityId: proofId._id,
      newValue: proofId,
      ipAddress: req.ip,
      userAgent: req.headers['user-agent']
    });

    res.send(proofId);
  } catch (error) {
    res.status(500).send(error);
  }
};

// Update a proof ID by ID
exports.updateProofId = async (req, res) => {
  const updates = Object.keys(req.body);
  const allowedUpdates = ['idType', 'idNumber', 'issuedBy', 'issueDate', 'expiryDate'];
  const isValidOperation = updates.every((update) => allowedUpdates.includes(update));

  if (!isValidOperation) {
    return res.status(400).send({ error: 'Invalid updates!' });
  }

  try {
    const proofId = await ProofId.findById(req.params.id);
    if (!proofId) {
      return res.status(404).send();
    }

    updates.forEach((update) => proofId[update] = req.body[update]);
    await proofId.save();

    // Create an audit log entry for the proof ID update
    await AuditLog.create({
      userId: req.user._id,
      action: 'update',
      entity: 'ProofId',
      entityId: proofId._id,
      oldValue: { ...proofId._doc }, // Store old values
      newValue: proofId,
      ipAddress: req.ip,
      userAgent: req.headers['user-agent']
    });

    res.send(proofId);
  } catch (error) {
    res.status(400).send(error);
  }
};

// Delete a proof ID by ID
exports.deleteProofId = async (req, res) => {
  try {
    const proofId = await ProofId.findByIdAndDelete(req.params.id);
    if (!proofId) {
      return res.status(404).send();
    }

    // Create an audit log entry for the proof ID deletion
    await AuditLog.create({
      userId: req.user._id,
      action: 'delete',
      entity: 'ProofId',
      entityId: proofId._id,
      oldValue: proofId,
      ipAddress: req.ip,
      userAgent: req.headers['user-agent']
    });

    res.send(proofId);
  } catch (error) {
    res.status(500).send(error);
  }
};

File: controllers/proof_id_controller.py
from fastapi import HTTPException, Request
from models import ProofId, AuditLog, User  # Assuming Beanie models
from typing import Dict, Any, List

class ProofIdController:
    @staticmethod
    async def create_proof_id(data: Dict[str, Any], request: Request, user_id: str) -> Dict[str, Any]:
        """Create a new proof ID and log the action."""
        try:
            data["user_id"] = user_id
            proof_id = ProofId(**data)
            await proof_id.insert()

            await AuditLog(
                user_id=user_id,
                action="create",
                entity="proofId",
                entity_id=str(proof_id.id),
                new_value=proof_id.dict(),
                ip_address=request.client.host,
                user_agent=request.headers.get("user-agent"),
            ).insert()

            return proof_id.dict()
        except Exception as e:
            raise HTTPException(status_code=400, detail=str(e))

    @staticmethod
    async def get_all_proof_ids(request: Request, user_id: str) -> Dict[str, Any]:
        """Retrieve all proof IDs and log the action."""
        try:
            proof_ids = await ProofId.find_all().populate("user_id", ["name", "email"]).to_list()
            proof_ids_dict = [proof_id.dict() for proof_id in proof_ids]

            await AuditLog(
                user_id=user_id,
                action="view",
                entity="proofId",
                entity_id=None,
                new_value=proof_ids_dict,
                ip_address=request.client.host,
                user_agent=request.headers.get("user-agent"),
            ).insert()

            return {"proofIds": proof_ids_dict}
        except Exception as e:
            raise HTTPException(status_code=500, detail=str(e))

    @staticmethod
    async def get_proof_id_by_id(proof_id: str, request: Request, user_id: str) -> Dict[str, Any]:
        """Retrieve a proof ID by ID and log the action."""
        try:
            proof_id_obj = await ProofId.get(proof_id)
            if not proof_id_obj:
                raise HTTPException(status_code=404, detail="Proof ID not found")

            await AuditLog(
                user_id=user_id,
                action="view",
                entity="proofId",
                entity_id=str(proof_id_obj.id),
                new_value=proof_id_obj.dict(),
                ip_address=request.client.host,
                user_agent=request.headers.get("user-agent"),
            ).insert()

            return proof_id_obj.dict()
        except Exception as e:
            raise HTTPException(status_code=500, detail=str(e))

    @staticmethod
    async def update_proof_id(proof_id: str, data: Dict[str, Any], request: Request, user_id: str) -> Dict[str, Any]:
        """Update a proof ID and log the action."""
        allowed_updates = {"id_type", "id_number", "issued_by", "issue_date", "expiry_date"}
        if not all(key in allowed_updates for key in data):
            raise HTTPException(status_code=400, detail="Invalid updates!")

        try:
            proof_id_obj = await ProofId.get(proof_id)
            if not proof_id_obj:
                raise HTTPException(status_code=404, detail="Proof ID not found")

            old_value = proof_id_obj.dict()
            for key, value in data.items():
                setattr(proof_id_obj, key, value)
            await proof_id_obj.save()

            await AuditLog(
                user_id=user_id,
                action="update",
                entity="proofId",
                entity_id=str(proof_id_obj.id),
                old_value=old_value,
                new_value=proof_id_obj.dict(),
                ip_address=request.client.host,
                user_agent=request.headers.get("user-agent"),
            ).insert()

            return proof_id_obj.dict()
        except Exception as e:
            raise HTTPException(status_code=400, detail=str(e))

    @staticmethod
    async def delete_proof_id(proof_id: str, request: Request, user_id: str) -> Dict[str, Any]:
        """Delete a proof ID and log the action."""
        try:
            proof_id_obj = await ProofId.get(proof_id)
            if not proof_id_obj:
                raise HTTPException(status_code=404, detail="Proof ID not found")

            old_value = proof_id_obj.dict()
            await proof_id_obj.delete()

            await AuditLog(
                user_id=user_id,
                action="delete",
                entity="proofId",
                entity_id=proof_id,
                old_value=old_value,
                ip_address=request.client.host,
                user_agent=request.headers.get("user-agent"),
            ).insert()

            return proof_id_obj.dict()
        except Exception as e:
            raise HTTPException(status_code=500, detail=str(e))